sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala

```diff
@@ -0,0 +1,334 @@
+   *                   between original and target calendar was changed,
+   *                   and differences in micros between calendars
+   * @param value The number of micros since the epoch 1970-01-01 to be rebased to the
+   *              target calendar.
+   * @return The rebased micros.
+   */
+  private def rebaseMicros(rebaseInfo: RebaseInfo, value: Long): Long = {
+    val switches = rebaseInfo.switches
+    var i = switches.length
+    do { i -= 1 } while (i > 0 && value < switches(i))
+    value + rebaseInfo.diffs(i)
+  }
+
+  // Loads rebasing info from an JSON file. JSON records in the files should conform to
+  // `JsonRebaseRecord`. AnyRefMap is used here instead of Scala's immutable map because
+  // it is 2 times faster in DateTimeRebaseBenchmark.
+  private[sql] def loadRebaseRecords(fileName: String): AnyRefMap[String, RebaseInfo] = {
+    val file = Thread.currentThread().getContextClassLoader.getResource(fileName)
+    val mapper = new ObjectMapper() with ScalaObjectMapper
+    mapper.registerModule(DefaultScalaModule)
+    val jsonRebaseRecords = mapper.readValue[Seq[JsonRebaseRecord]](file)
+    val anyRefMap = new AnyRefMap[String, RebaseInfo]((3 * jsonRebaseRecords.size) / 2)
+    jsonRebaseRecords.foreach { jsonRecord =>
+      val rebaseInfo = RebaseInfo(jsonRecord.switches, jsonRecord.diffs)
+      var i = 0
+      while (i < rebaseInfo.switches.length) {
+        rebaseInfo.switches(i) = rebaseInfo.switches(i) * MICROS_PER_SECOND
+        rebaseInfo.diffs(i) = rebaseInfo.diffs(i) * MICROS_PER_SECOND
+        i += 1
+      }
+      anyRefMap.update(jsonRecord.tz, rebaseInfo)
+    }
+    anyRefMap
+  }
+
+  /**
+   * A map of time zone IDs to its ordered time points (instants in microseconds since the epoch)
+   * when the difference between 2 instances associated with the same local timestamp in
+   * Proleptic Gregorian and the hybrid calendar was changed, and to the diff at the index `i` is
+   * applicable for all microseconds in the time interval:
+   *   [gregJulianDiffSwitchMicros(i), gregJulianDiffSwitchMicros(i+1))
+   */
+  private val gregJulianRebaseMap = loadRebaseRecords("gregorian-julian-rebase-micros.json")
+
+  /**
+   * Rebases the given `micros` to the number of microseconds since the epoch via a local
+   * timestamp that have the same date-time fields in Proleptic Gregorian and in the hybrid
+   * calendars.
+   *
+   * The function may optimize the rebasing by using pre-calculated rebasing maps. If the maps
+   * don't contains information about the current JVM system time zone, the functions falls back
+   * to regular conversion mechanism via building local timestamps.
+   *
+   * Note: The function assumes that the input micros was derived from a local timestamp
+   *       at the default system JVM time zone in Proleptic Gregorian calendar.
+   *
+   * @param micros The microseconds since the epoch 1970-01-01T00:00:00.000000Z.
```

Comment:
`The microseconds .... in Julian calendar.`

Reply:
changed